\documentclass[11pt]{article}

\usepackage{fullpage,times,namedplus,pgf}
\usepackage{enumitem}
\usepackage{amsmath,amssymb}
\usepackage{kbordermatrix}% http://www.hss.caltech.edu/~kcb/TeX/kbordermatrix.sty

\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator{\E}{\mathbb{E}}


\usepackage{algorithm}
\usepackage[noend]{algorithmic}
\renewcommand{\algorithmicrequire}{\textbf{Input:}}
\renewcommand{\algorithmicensure}{\textbf{Output:}}

\newcommand{\myroot}{\ensuremath{\mathrm{root}}}

\newcommand{\context}{\ensuremath{\mathrm{con}}}
\newcommand{\psn}{\ensuremath{\mathrm{pos}}}

\newcommand{\birth}{\ensuremath{\mathcal{B}}}
\newcommand{\death}{\ensuremath{\mathcal{D}}}
\newcommand{\expansion}{\ensuremath{\mathcal{E}}}
\newcommand{\contraction}{\ensuremath{\mathcal{C}}}
\newcommand{\merging}{\ensuremath{\mathcal{M}}}
\newcommand{\separating}{\ensuremath{\mathcal{S}}}
\newcommand{\child}[1]{\ensuremath{\mathrm{child}(#1)}}


\title{A model of epigenome evolution}
\author{Jianghan Qu \and Andrew D. Smith}

%%% ADS Notes:
%%
%%% (1) We need to define operationally an epigenome as a binary
%%% sequence, and in the introduction we need to explain this.
%%
%%% (2) We need to make sure that we use the terms ``position'' and
%%% ``site'' consistently.

\begin{document}

\maketitle

\begin{abstract}
  Epigenetic marks along the mammalian genome are organized into
  alternating genomic domains bearing and lacking the mark. The
  location and size of domains enriched for an epigenetic mark are
  indicative of the presence, function and activity of regulatory
  elements and the chromatin states. Comparative epigenomic studies
  aim to resolve the evolutionary history of regulatory elements by
  comparing epigenomic profiles in multiple species.  However,
  computational methods for comparing epigenetic marks at high
  resolution, inferring evolution rates along different phylogenetic
  lineages and reconstructing the evolutionary history are still
  limited.  In this study, we aim to establish a simulation, sampling
  and inference framework for studying the evolution of the genomic
  distribution of an epigenetic from the profiles of multiple extant
  species.  We model the profile of an epigenetic mark in a species
  with a two-state Markov chain, and model the evolution of an
  epigenomic sequence with a continuous-time Markov chain, where
  instantaneous transition rates at a site is dependent on the
  contemporary states of its neighboring sites. We use a MCMC sampling
  method for estimating the context-dependent transition rates and
  inferring the evolutionary history that lead to diverse profiles in
  extant species from a common ancestral epigenome.  We show with
  applications to DNA methylation and histone modification profiles
  that our methods can reveal both genome-wide evolutionary features
  through estimates of the model parameters and high-resolution
  evolutionary patterns in local regions through posterior sampling of
  the evolutionary history.
\end{abstract}

\section{Introduction}

%%% First paragraph intro about epigenomes
The epigenome of a mammalian cell reflects much of the complexity we
associate with cell phenotype and behaviors \cite{}. Individual
epigenomic marks, for example a histone modification, may be viewed
from a simpler perspective as contiguous genomic intervals where the
presence or absense of that mark is associated with genomic
function. Intervals of the genome that have a high density of H3K9me3
are often associated with condensed chromatin state and silencing of
genes within those intervals
\cite{nakayama2001role}. Genomic intervals with high
density of H3K4me3, on the other hand, are associated with
accessibility by transcription factors and are associated with gene
promoters \cite{santos2002active}. So despite the complexity often
ascribed to the mammalian epigenome \cite{bernstein2007mammalian},
studies focusing on individual epigenomic modifications have been
highly successful in elucidating transcriptional regulation in a
variety of systems \cite{martin2005diverse}.

%%% Need to cite work that has boiled down the complexity of
%%% epigenomes to simpler ``functional'' states, like ``open''
%%% ``poised, silenced, ``actively transcribed'', etc.

%%% Need to describe birth, death, etc.

One challenge in modeling epigenome evolution is that desirable models
should account for the inherent auto-correlation of epigenomic state
along the genome. The most well known models of molecular evolution,
applied to amino acids or nucleotides, treat each site as evolving
independently -- a simplifying assumption that has proven very useful.
When models allow for dependencies between sites, we additionally hope
that those dependencies can be interpreted.

%%% Summary and timeline of existing work
\citetext{pedersen1998codon} examined the problem of modeling
evolution at the codon level and designed an approach capable of
describing CpG depression across codon boundaries, a form of
interdependence between adjacent codons. With similar motivation,
\citetext{jensen2000probabilistic} examined the properties of
evoluationary models for which the stationary distribution on
sequences naturally exhibits particular frequencies of dinucleotides.
The result was an approach to model the evolution of an individual
nucleotide as a function of that nucleotide's neighbors in a way that
induces a Markov process on the stationary distribution.
%%% Need to talk about the sufficient condition given by J-P (2000)

%%% Need to relate our work back to the phylo-HMM stuff, particularly
%%% because that work was aware of the chain graph perspective, and
%%% our ``states'' are defined analogously to the generalization from
%%% the Koller-Friedman book.
Our goal of modeling the evolution of an epigenome can be viewed in
analogy with phylogenetic hidden Markov models \cite{}. The phastCons
algorithm \cite{} is the best known variant of phylogenetic HMM, and
has dramatically impacting the field of comparative genomics by
providing a general approach to model ``conserved'' genomic intervals
for a set of species. The states of a simple phylogenetic HMM
correspond to alternating conserved and non-conserved intervals of
aligned genomes. The generalization associates a binary state label
with each nucleotide in each species and presents algorithmic
challenges when viewed generally as a chain graph
\cite{koller2009probabilistic}.

The remainder of this paper is organized as follows. We first describe
our model for epigenome evolution as an adaptation of the principles
introduced by \citetext{jensen2000probabilistic} and use simulation fo
explore model parameterization. Then we explain how to inferences are
made in the context of this model, using simulation to demonstrate the
accuracy of our procedures. Finally, we apply this model on an
existing data set.

\section{The model}

\subsection{Biological assumptions and assumed representation for the epigenome}
\label{biodefs}

%%% Need to cover the issues here of what it means for a methylome,
%%% what it means for ATAC, what it means for histone mods.
We assume that the epigenome is a sequence of binary states
super-imposed on the genome. This assumption is restrictive, but it
allows us to consider the epigenome from either the perspective of an
individual epigenomic modification, or as reflecting a particular type
of functional interval. One example of former is a sequence of binary
variables corresponding to the presence or absence of a H3K27me3.
Another example is a binary variable to indicate accessibility, as
determined by a particular assay. An example of a functionally-defined
binary variable may be ``accessible'' or ``enhancer,'' both of which
can be associated with different epigenomic modifications, but are
known to be organized as contiguous intervals. Epigenomic state is
correlated along the genome as a reflection of the organization of the
epigenome into contiguous intervals.

Most types of data that inform us about the epigenome are based on
sequencing, and usually based on density of mapped reads (e.g. from
ChIP-seq or ATAC-seq). Often these data are summarized in
non-overlapping ``bins'' through the genome. When we refer to a
position in the epigenome, we assume that such position corresponds in
a meaningful way to either individual nucleotide positions
in the genome, or appropriate bins. We only require that the
neighboring relations are preserved.

Let epigenome $s$ be the sequence $s=s_1s_2\cdots s_N$ with $s_i$
denoting the state at position $i$.
%%%
As a graphical model, neighboring sites are connected with undirected
edges. An epigenome $s$ evolves over time, but we assume the
stationary distribution for the epigenome has a Gibbs distribution
that factorizes over pairs of neighboring sites:
\begin{equation}\label{eqn:stationary}
  \Pr(s) = \frac{1}{Z} \exp\big\{\phi(s_1) +\sum_{n=1}^{N-1}\phi(s_n, s_{n+1}) + \phi(s_N)\big\}
\end{equation}

As the epigenome evolves, the state associated with individual
positions may change. The types of changes we are most interested in
can be more naturally interpreted in terms of contiguous intervals. We
use the term epigenomic ``feature'' to refer to a consecutive interval
having the ``1'' state. The main types of changes we are interested in
are the following:
\begin{itemize}
\item {\it Birth and death:} During evolution, some new epigenomic
  feature may appear, or a feature that existed in an ancestor may
  disappear. %%% Should show examples
\item {\it Expansion and contraction:} As the epigenome evolves,
  epigenomic segments can become wider or more narrow, and this may
  happen in either direction.
\item {\it Merging and separating:} Two epigenomic features that are
  nearby in the ancestral genome may merge into a single interval.
  Conversely, a single epigenomic feature in the ancestral epigenome
  may separate into two intervals.
\end{itemize}
We will describe a model that treats the expansion and contraction of
features symmetrically in both directions along the genome.  This
choice in modeling is often reasonable, but not always. For example,
certain epigenomic modifications are frequently associated with parts
of genes (e.g. promoters), and the genes themselves have
directionality.
%%% Need to give examples where we believe this is reasonable, like
%%% enhancers that operate at a distance. But then cite CTCF
%%% directionality.

%%% Describe the paths for the full genome, and then the paths for
%%% individual sites
We are also interested in modeling sequences that evolve in continuous
time, and we use two representations to describe a particular
realization of the evolutionary process. These two representations
have distinct advantages, both in explaining aspects of our model and
in different computations.

Let $s$ be a sequence of $N$ binary states that evolve according to a
continuous time Markov process. The state space for $s$ has size
$2^N$, but we constrain instantaneous transition rates such that each
change to $s$ only involves a single site (coordinate) within $s$. We
can summarize the process as a {\it global path} over time interval
$[0,\tau)$ as follows:
\[
\begin{array}{lll}
H = (s, Y, V), ~\mbox{ where} & s \in \{0,1\}^N,\\
                             & Y = (y_1, \ldots, y_w),\\
                             & 0\leq y_i < y_{j} < \tau, & \mbox{ for } i < j,\\
                             & V = (v_1,\ldots, v_w), & \mbox{ for } 1\leq v_i\leq n.
\end{array}
\]
For any time $0 \leq t < \tau$, we let $s(t)$ denote the state
sequence after having applied changes to $s = s(0)$ at positions given
by $v_1,v_2,\ldots,v_k$ such that $y_k$ is maximal satisfying $y_k
\leq t$. From $H$ we can obtain $s(t)$ for any $0\leq t < \tau$ by
tracing changes made to $s$.
%%%

We can organize the same information differently and describe the
process in terms of {\it site-specific paths}. In particular, for site
$1\leq n \leq N$,
\[
\begin{array}{lll}
H_n = (s_n, Y_n), ~\mbox{ where} & s_n \in \{0,1\},\\
                                & Y_n = (y_1, \ldots, y_{w_n}),\\
                                & 0\leq y_i < y_{j} < \tau, & \mbox{ for } i < j.
\end{array}
\]
Our assumption that instantaneous rates are non-zero only for
transitions that modify a single position in $s$ defines bijection
between the global jump times $Y$ and the union of the site-specific
jump times $\cup_n Y_n$. Note that the sequences of site-specific
paths $(H_1,\ldots,H_N)$ more concisely describes a realization of our
process; obtaining the state sequence $s(t)$ for any time $t$ requires
more effort.

%%% Introduce transition rates
The evolution of states at a given site in the state sequence $s$
depends on the states of its neighboring sites at all times during
evolution. Consider site $n$ and its two immediate neighbors $n-1$ and
$n+1$, and let the contemporaneous states for these sites at time $t$
be $s_{n-1}(t) = i$, $s_n(t)=j$ and $s_{n+1}=j$. The instantaneous
rate for a transition at site $n$ from state $j$ to the complementary
binary state $\bar{j}$ is $\lambda(i, j, k)$. In other words, the
instantaneous rate $\lambda(i, j, k)$ is a function of the states at
the site of interest and its two immediate neighbors.  Since we assume
the evolutionary process is symmetric with respect to the direction
along the genome, we require $\lambda(i, j, k) = \lambda(k, j, i)$.
%%% ADS: note sure what this means below:
When the states at all positions except for site $n$ remain constant
over time interval the state $s_n$ follows a two-state continuous-time
Markov process within that time interval.

\subsection{Stationary Gibbs measure as Markov chain}

The stationary distribution in \eqref{eqn:stationary} is equivalent to
the distribution of a Markov chain. Therefore, when the epigenome is
modeled with a Gibbs measure, we can sample an instance of the state
sequence, or evaluate its probability, using the equivalent Markov
chain. We can relate the factors in \eqref{eqn:stationary} and the
transition probabilities of the Markov chain as follows. Define
pair-wise potentials $Q(i,j)=\exp(\phi(i, j))$, where $i, j\in\{0,1\}$
are binary states. The largest eigenvalue of $Q$ is
\[
q=\textstyle\frac{1}{2}\left(Q(0,0)+Q(1,1) +\sqrt{\Delta}\right), \text{ where }
\Delta=(Q(0,0) - Q(1,1))^2 + 4Q(0,1)Q(1,0).
\]
Let $h$ be a right eigenvector of $Q$ corresponding to $q$, then we have
\[
\frac{h_0}{h_1} = \frac{Q(0,1)}{q-Q(0,0)} = \frac{q-Q(1,1)}{Q(1,0)}.
\]
The Markov chain transition matrix is:
\[
T(i,j) = \frac{Q(i,j)h_j}{qh_{i}},~\text{where } i,j \in\{0,1\}.
\]
More specifically,
\begin{equation} \label{eqn:gibbs2markov}
  \begin{array}{ll}
    T(1,1) = \displaystyle\frac{2Q(1,1)}{Q(0,0)+Q(1,1)+\sqrt{\Delta}}, &
    T(0,0) = \displaystyle\frac{2Q(0,0)}{Q(0,0)+Q(1,1)+\sqrt{\Delta}}, \\[2em]
    T(0,1) = \displaystyle\frac{4Q(0,1)Q(1,0)}{(Q(0,0)+\sqrt{\Delta})^2 -Q(1,1)^2}, &
    T(1,0) = \displaystyle\frac{4Q(0,1)Q(1,0)}{(Q(1,1)+\sqrt{\Delta})^2 -Q(0,0)^2}.
  \end{array}
\end{equation}
Accordingly, the expected fraction of an epigenome state sequence
residing within functional domains (the ``features'') is:
\[
1- \frac{2Q(0,1)Q(1,0)}{(Q(0,0)-Q(1,1))^2 + 4Q(0,1)Q(1,0) +
  (Q(1,1)-Q(0,0))\sqrt{\Delta}}.
\]

\subsection{Relating the substitution model and the stationary distribution}

\citetext{jensen2000probabilistic} gave a sufficient condition for a
continuous time evolutionary model with the properties we outlined for
$\lambda$ to have a stationary distribution of the form $\phi$. In
particular, if
%%% Prop 1 of J-P 2000
\begin{equation}\label{eqn:prop1}
  \frac{\lambda_{ijk}}{\lambda_{i\bar{j}k}} =
  \frac{\lambda(i, j, k)}{\lambda(i, \bar{j}, k)} =
  \frac{\exp(\phi(i, \bar{j})+ \phi(\bar{j}, k))}{\exp(\phi(i, j)+ \phi(j, k))},
\end{equation}
then \eqref{eqn:stationary} is the stationary distribution for the
Markov process with instantaneous rates $\lambda_{ijk}$.

\citetext{jensen2000probabilistic} also provide (Proposition 2) a way
of specifying $\lambda$ from $\phi$: substitution rates $\lambda$
satisfy the relation \eqref{eqn:prop1} if and only if they can be
written in the form
\begin{equation}\label{eqn:prop1}
  \log(\lambda_{ijk}) = -\psi(i, j, k) + \ell(i, k).
\end{equation}
This criteria was introduced in the context of an arbitrary number of
states for each position in the state sequence, and the $\ell$
function can be understood as $\ell(j, j'; i, k)$, for any two
different states $j$ and $j'$, which is symmetric in $(j, j')$. In our
setting the states are binary and $\ell$ is only a function of the two
neighboring states $(i,k)$. Moreover, we can directly verify that if
we define substitution rates as
\begin{equation}
  \log (\lambda_{ijk}) =  \ell(i, k) + (\phi(i, \bar{j}) +\phi(\bar{j}, k)),
\end{equation}
then they satisfy the condition of \eqref{eqn:prop1}.

\subsection{Parameterization and interpretation}

We can organize the neighbor-dependent transition rates of defined
above according to an $8\times8$ matrix:
\begin{equation}\label{eqn:Lambda}
  \renewcommand{\kbldelim}{(}% Left delimiter
  \renewcommand{\kbrdelim}{)}% Right delimiter
  \Lambda = \kbordermatrix{
    & 000 & 010 & 001 & 011 & 100 & 110 & 101 & 111 \\
    000 & \cdot & \mathcal{B} & 0 & 0 & 0 & 0 & 0 & 0 \\
    010 & \mathcal{D} & \cdot & 0 & 0 & 0 & 0 & 0 & 0 \\
    001 & 0 & 0 & \cdot & \mathcal{E} & 0 & 0 & 0 & 0 \\
    011 & 0 & 0 & \mathcal{C} & \cdot & 0 & 0 & 0 & 0 \\
    100 & 0 & 0 & 0 & 0 & \cdot & \mathcal{E} & 0 & 0 \\
    110 & 0 & 0 & 0 & 0 & \mathcal{C} & \cdot & 0 & 0 \\
    101 & 0 & 0 & 0 & 0 & 0 & 0 & \cdot & \mathcal{M} \\
    111 & 0 & 0 & 0 & 0 & 0 & 0 & \mathcal{S} & \cdot
  }
\end{equation}
We may interpret the non-zero entries in $\Lambda$ as corresponding to
biological events outlined in Section~\ref{biodefs}. The values
$\mathcal{B}$ and $\mathcal{D}$ are the rates of ``birth''
and ``death,'' respectively, for epigenomic features. The value
$\mathcal{M}$ corresponds to the merging of two features into a single
contiguous interval ($101\rightarrow 111$). Conversely, the value
$\mathcal{S}$ corresponds to epigenomic features ``splitting'' and
becoming two separate intervals ($111\rightarrow 101$). The remaining
non-zero values, $\mathcal{E}$ and $\mathcal{C}$, correspond to the
expansion (widening) and contraction (narrowing), of epigenomic
features. Both of these parameters appear twice in $\Lambda$,
reflecting our assumption that the rates governing any widening or
narrowing of intervals do not depend on direction, as explained in
Section~\ref{biodefs}.

If an epigenome evolves according to substitution rates $\Lambda$ with
stationary distribution \eqref{eqn:stationary}, then the condition in
\eqref{eqn:prop1} holds, and we have the following constraints for the
rates in the above matrix:
\begin{equation}\label{eqn:constraint}
  \mathcal{B}\mathcal{C}^2\mathcal{M}=\mathcal{D}\mathcal{E}^2\mathcal{S}.
\end{equation}
So given substitution rates $\birth{}, \death{}, \expansion{},
\contraction{}$, we have the following relationships between
horizontal potentials:
\begin{equation}\label{eqn:rel}
  \phi(0,0) = \phi(0,1) +\frac{1}{2}\log\left(\frac{\death{}}{\birth{}}\right), ~\text{ and }~
  \phi(1,1) = \phi(0,1) +\frac{1}{2}\log\left(\frac{\death{}\expansion{}^2}{\birth{}\contraction{}^2}\right).
\end{equation}

In summary, the transition rate matrix $\Lambda$ has 5 free
parameters. If we add an additional constraint on the expected number
of changes per unit time, then the model will have only 4 free
parameters. The two ratios $\death{}/\birth{}$ and
$\death{}\expansion{}^2/(\birth{}\contraction{}^2)$ then uniquely
determine the stationary distribution \eqref{eqn:stationary} through
equation \eqref{eqn:rel}.

\section{Simulate epigenomic changes during evolution}

%%% ADS: do we need something in here to mention that we verified that
%%% simulation does work?

We model the evolution of the entire epigenome (as a sequence of
states) using a continuous time Markov process that only allows
instantaneous transitions from one sequence to another if the two
differ at a single position. The jumping rate at each position in the
epigenome is dependent on the state at that position and on the states
of the left and right neighboring positions. We assume for convenience
that the states of the first and last sites are fixed throughout
evolution.

Consider the $2^N \times 2^N$ transition rate matrix $R$ for any pair
of epigenomes $x$ and $x'$ that differ at exactly one position. The
state at that position is $j$ in epigenome $x$ and $\bar{j}$ in
epigenome $x'$. Neighboring positions in both $x$ and $x'$ have states
$i$ and $k$. The instantaneous rate of a jump between $x$ and $x'$ is
$\lambda_{ijk}$. The holding time in a state sequence $x$ is
exponentially distributed with rate $-R_{xx}$ equal to the sum of
instantaneous rates for jumps from $x$ to any state sequence $x'$ that
differs from $x$ at exactly one position. This rate for $x$ can be
expressed in terms of rates on triplets:
\[
-R_{xx} = \sum_{i,j,k}c_{ijk}(x)\lambda_{ijk},
\]
where $c_{ijk}(x) = \sum_{n=1}^{N-2}I(x_{n}=i, x_{n+1}=j, x_{n+2}=k)$
counts instances of the triplet pattern $ijk$ in $x$.

Given that a jump has occurred, the probability that the jump changed
$x$ at the middle position of triplet $ijk$ is proportional to
$c_{ijk}(x)\lambda_{ijk}$. Further, given that a jump occurred with
context $ijk$, our model assumes the jump is equally likely to have
changed any position in $x$ having state $j$ with left and right
neighbors having states $i$ and $k$.
%%
The expected number of changes per site, per unit time, is
%%% ADS: use of pi
$\sum_{ijk}\pi_{ijk}\lambda_{ijk}$, where $\pi_{ijk}$ is the
stationary distribution for the pattern $ijk$ in the epigenome.
%%% ADS: above, can we generalize away from epigenome here?
%%
These assumptions suggest a simulation procedure in Algorithmfor the evolutionary
process followed by an epigenome of $N$ sites over a time interval
$[0, \tau)$, starting from an initial sequence $s$.

\begin{algorithm}[t]
  \begin{algorithmic}[1]
    \caption{Simulating epigenome evolution}
    \REQUIRE Binary state sequence $s$ of length $N$, time $\tau$ and rates $\Lambda$.
    \ENSURE Simulated global path $H = (s, Y, V)$ for change times $Y$ and positions $V$.
    \STATE $t \gets 0$, $V\gets\emptyset$, $Y\gets\emptyset$ and $x \gets s$
    \WHILE {$t < \tau$}
    \STATE Sample holding time $y\sim \mathit{Exp}(-R_{ss})$, where $-R_{ss} = \sum_{i,j,k}c_{ijk}(x)\lambda_{ijk}$
    \STATE $t \gets t + y$
    \IF {$t < \tau$}
    \STATE Sample binary triplet $ijk$ with probability proportional to $c_{ijk}(x)\lambda_{ijk}$
    \STATE Sample position $n$ uniformly from the $c_{ijk}(x)$ at center of a $ijk$ triplet in $x$
    \STATE Modify state sequence $x$ to have state $\bar{j}$ at position $n$
    \STATE Append $n$ to $V$ and append $t$ to $Y$
    \ENDIF
    \ENDWHILE
    %% \STATE Append a copy of current state sequence $x$ to $X$ and append time $\tau$ to $Y$
    \RETURN $H = (s, Y, V)$
  \end{algorithmic}
\end{algorithm}

%%% Implementation!!!

%%% plans for figures:
%%% multiple parameter settings (stationary, non-stationary, different rates)
%%% (1) distribution of patterns (stays stable, or moves towards target stationary state)
%%% (2) simulated paths with different parameter, show examples and summary statistics for events

\section{Parameter estimation}

If we are given the complete epigenome evolution path from time $0$ to
time $t$, can we effectively recover the initial distribution and
mutation parameters and evolutionary time? We are interested in the
parameters describing the evolutionary process, which are the
transition rates $\{\lambda_{ijk}\}$. These parameters will be
inferred from the state changes in the evolutionary path for the
entire epigenome. Meanwhile, we do not require the process to be
stationary, so we may also interested in the properties of the
epigenome at time 0, which are characterized by the Markov chain
transition probabilities $T_{0}$. These transition probabilities are
easy to infer given the complete observations at the time-0 epigenome.

Let $c_{ij} = \sum_{n=1}^{N-1}I\{s_n(0) =i, s_{n+1}(0)=j\}$. Then
\[
\hat{T}_{00} = \frac{c_{00}}{\sum_{n=1}^{N-1}I\{s_n(0) = 0\}}, ~
\hat{T}_{11} = \frac{c_{11}}{\sum_{n=1}^{N-1}I\{s_n(0) = 1\}}.
\]

We first assume that the time span of this complete evolutionary
history is known, \textit{i.e.} the value of $t$ is given.

Recall that $L_n = \{s_n(0), K, \{t_k\}_{k=1}^K, t\}$ is a full path
at position $n$ in the epigenome. We can pool all the jumping times at
all positions as an ordered sequence of timepoints $J = \{(t_m,
\psn{}_m, \context{}_m\}_{m=1}^{M}$, where $\psn{}_m$ is the position
of the $m$-th jump in the entire evolutionary history of the
epigenome, $\context{}_m$ is the 3-tuple context of the change that
occurred at the $m$-th jump.

Let $\Delta_m = t_{m} - t_{m-1}$ be the holding time just prior to the
$m$-th jump for the evolving state sequence. Using $c_{ijk}(t_{m-1})$ to
denote the triplet distribution for the state sequence at time
$t_{m-1}$, the holding time $\Delta_m$ follows the exponential
distribution
\[
\Delta_m \sim \mathit{Exp}(\lambda_m), ~\text{ with }~
\lambda_{m} = \sum_{i,j,k}c_{ijk}(t_{m-1})\lambda_{ijk}.
\]

The likelihood function for parameters $\{\lambda_{ijk}\}$ is thus
\begin{equation}\label{eqn:lik}
  \prod_{m=1}^{M} \lambda_{m} \exp(-\lambda_{m}\Delta_m) \times \frac{\lambda_{\context{}_m}}{\lambda_m}
  =\prod_{m=1}^{M}\lambda_{\context{}_m}\exp(-\lambda_m\Delta_m).
\end{equation}
%%
And the log-likelihood function is
\begin{equation}\label{eqn:loglik1}
    l = \sum_{i,j,k} \left(\sum_{m=1}^M\log\lambda_{ijk}\times I_{\{\context{}_m = ijk\}} - c_{ijk}(t_m^-)\lambda_{ijk}\Delta_m\right)
     = \sum_{i,j,k} \big(J_{ijk}\log\lambda_{ijk} - D_{ijk}\lambda_{ijk} \big)
\end{equation}
where
\[
J_{ijk} = \sum_{m=1}^M I\{\context{}_m = ijk\}
~~\mbox{ and }~~
D_{ijk} = \sum_{m=1}^Mc_{ijk}(t_{m-1})\Delta_m.
\]
In other words, $J_{ijk}$ counts the number of jumps creating the
triplet $ijk$ and $D_{ijk}$ is the time spent in context $ijk$
aggregated over all sites.

%% where $J_{ijk} = \sum_{m=1}^M I_{\{\context{}_m = ijk\}}$, and
%% \begin{equation}\label{eqn:Dijk}
%%   \begin{aligned}
%%     D_{ijk} &= \sum_{m=1}^Mc_{ijk}(t_m^-)\Delta_m \\
%%     & = \sum_{m=1}^M \bigg(\sum_{n=1}^{N} I_{\{\text{con}(n; t_m^-) = ijk\}}\bigg)\times(t_m - t_{m-1}) \\
%%     & = \sum_{n=1}^{N}\int_{0}^{t_M} I_{\{\text{con}(n;t) = ijk\}} dt\\
%%     & = \text{Total time in context $ijk$ aggregated over all sites}
%%   \end{aligned}
%% \end{equation}
%% where $\text{con}(n; t)$ is the sequence context of site $n$ at time $t$.

The constraints on the transition rates $\lambda_{ijk}$ as indicated
in the matrix \eqref{eqn:Lambda} and equation \eqref{eqn:constraint}
are:
\begin{equation}\label{eqn:constraints}
  %% \left\{
  \begin{array}{c}
    \lambda_{001} = \lambda_{100}\\
    \lambda_{011} = \lambda_{110}\\
    \lambda_{000}\lambda_{110}^2\lambda_{101} = \lambda_{010}\lambda_{100}^2\lambda_{111}
  \end{array}
  %% \right.
\end{equation}
So for the purpose of estimating parameters, we only require 5
gradients (see Appendix).
%%% ADS: above, this makes no sense if we use 4 independent parameters...?

%% \begin{equation}\label{eqn:loglik2}
%%   \begin{aligned}
%%     l = ~ & (J_{000}\log\lambda_{000} - D_{000}\lambda_{000}) ~ +
%%     (J_{010}\log\lambda_{010} - D_{010}\lambda_{010}) ~ +
%%     (J_{101}\log\lambda_{101} - D_{101}\lambda_{101}) ~ + \\
%%     & \big( (J_{100} + J_{001})\log\lambda_{001} - (D_{100}+D_{001})\lambda_{001}\big) + \\
%%     & \big((J_{011} + J_{110})\log\lambda_{011} - (D_{011}+D_{110})\lambda_{011}\big) +  \\
%%     & J_{111}\log\left(\frac{\lambda_{000}\lambda_{011}^2\lambda_{101}}{\lambda_{010}\lambda_{001}^2}\right) - D_{111}\frac{\lambda_{000}\lambda_{011}^2\lambda_{101}}{\lambda_{010}\lambda_{001}^2} \\
%%     = ~ & \sum_{c=0,2,5}\bigg(J_c\log\lambda_c - D_c\lambda_c\bigg) ~ +
%%     \sum_{c=1,3} \bigg((J_c + J_{c'})\log\lambda_c - (D_c + D_{c'})\lambda_c\bigg) ~ + \\
%%     ~ & \bigg(J_7\log(\frac{\lambda_0\lambda_3^2\lambda_5}{\lambda_2\lambda_1^2})
%%     - D_7\cdot\frac{\lambda_0\lambda_3^2\lambda_5}{\lambda_2\lambda_1^2} \bigg),
%%   \end{aligned}
%% \end{equation}
%% where $c'$ is the symmetric context pattern of $c$, \textit{i.e.} if
%% $c$ is $001$, then $c'$ is $100$.

\subsubsection{Estimates when evolutionary time is a parameter}
%%% ADS: I feel this is stated backwards. The paramters are not
%%% identifiable until we make the assumption about unit branch
%%% lengths.
In the context of phylogenetic inference, we are given the
observations at extant species, \textit{i.e.} leaf nodes of the
phylogenetic tree, and will rely on EM procedures to both estimate
model parameters (M-step) and making inferences about evolutionary
paths form the common ancestor (E-step).  In the E-step, we compute
the expected values of the complete-data sufficient statistics $J$s
and $D$s along individual branches conditional on the observed leaf
data and our current model parameters (branch lengths and transition
rates). In M-step, we find the set of parameters that give the best
likelihood with the expected values of sufficient
statistics. For an Exponential random variable
$X$:
\[
X \sim \mathit{Exp}(\lambda) \Leftrightarrow X/t \sim \mathit{Exp}(\lambda t),
~~ \text{for any constant $t>0$.}
\]
Let our current estimates of branch lengths be $\{\ell_b\}$, the new
branch lengths be $\{\ell'_b\}$, and the new rates be
$\{\lambda'_c\}$. Then for a jumping interval $\Delta_{bm}$ on branch
$b$ observed when the branch length is $\ell_b$, the scaled interval
length is $\Delta_{bm}\times \frac{\ell'_b}{\ell_b}$ under the new
branch length. Therefore, under the new model parameter set
$\{\ell'_b\}$ and $\{\lambda'_c\}$, we have
\[
\Delta_{mb}\times \frac{\ell'_b}{\ell_b} \sim \mathit{Exp}(\lambda'_{(bm)}) \Rightarrow
\Delta_{mb} \sim \mathit{Exp}(\lambda'_{(bm)} \times \frac{\ell'_b}{\ell_b}).
\]
%%%%%
Let $\tau_b = \frac{\ell'_b}{\ell_b}$ be the scaling factor for the
new branch lengths relative to the old branch lengths. Now we can
write the likelihood of observing all of the given jumping intervals
$\{\Delta_{b,m}\}$ under the new parameter set as :
\begin{equation}\label{eqn:liknew}
  \begin{aligned}
    L(\{\Delta_{b,m}\}; \{\ell'_b\}, \{\lambda'_c\})
    = & \prod_{b=1}^B\Bigg(\prod_{m=1}^{M_b-1}
    \lambda'_{(b,m)}\tau_b
    \exp(-\lambda'_{(b,m)}\tau_b\Delta_m) \times \frac{\lambda'_{\context{}_m}}{\lambda'_{(m)}}
    \Bigg) \times \exp(-\lambda'_{(b,M_b)}\tau_b\Delta_{M_b})\\
    = & \prod_{b=1}^B\Bigg(
    \prod_{m=1}^{M_b-1}\lambda'_{\context{}_m}\tau_b\exp(-\lambda'_{(m)}\tau_b\Delta_m)
    \Bigg)\times \exp(-\lambda'_{(b,M_b)}\tau_b\Delta_{M_b}).
  \end{aligned}
\end{equation}
%%%%%
The log likelihood becomes
\begin{equation}\label{eqn:logliknew}
  \begin{aligned}
    l = & \sum_{b=1}^B\sum_{m=1}^{M_b-1} \log(\lambda'_{c_m}\tau_b) -  \sum_{b=1}^B\sum_{m=1}^{M} \lambda'_{(m)}\Delta_{bm}\tau_b \\
    = & \sum_{b=1}^B\sum_{c} J_{bc}\log(\lambda'_c\tau_b)  - \sum_{b=1}^B\sum_{c} D_{b,c}\lambda'_c\tau_b ,
  \end{aligned}
\end{equation}
where $J_{b,c} = \sum_{m=1}^{M_b-1}I_{\{c_{b,m}= c\}}$, $D_{bc}
= \sum_{m=1}^{M_b}I_{\{c_{b,m} =c\}}\Delta_{b,m}$, $c_{b,m}$ is
the context of the $m$-th jump on branch $b$, and $\Delta_{bm}$ is the
holding time before the $m$-th jump on branch $b$ of length
$\ell_b$. Then, we can optimize \eqref{eqn:logliknew} over
$\{\tau_b\}$ and $\{\lambda'_c\}$, and set new branch lengths
$\{\ell'_b = \tau_b\ell_b\}$.

\subsection{Posterior distribution of a path given two neighboring paths}

Here we consider inference on the path for an single position in an
epigenome that is evolving for some specified time. We assume the
paths for all other positions are known for the entire evolutionary
time span. In particular, we will make inferences about site $n$ and
have access to known jump times $\{(t_m, \psn{}_m, \context{}_m)
\}_{m=1}^{M}$ for all positions except $n$.
%%
We also assume that the model parameters $\Lambda$ and total
evolutionary time $t$ are known, along with the the initial state of
the epigenome (i.e. the root). We seek inferences related to the
posterior distribution of the path $L_n$:
\[
\Pr(L_n|L_{-n}, \{\lambda_{ijk}\}) \propto \Pr(L_n\cup L_{-n}),
\]
where $L_{-n}$ denotes all paths for sites other than $n$. The
approach we adopt is using MCMC to sample from the posterior
distribution of $L_n$.

\section{Inferences in the context of a tree structure with fixed leaf data}

Now we assume the epigenome has evolved according to a tree structure,
and we wish to make inferences about a bifurcating evolutionary
process. The paths associated with a given site must satisfy
additional constraints to be ``consistent.'' We first consider a tree
with 3 nodes, a root $u$ with a single child node $v$ and two leaf
nodes below $v$, denoted $w_1$ and $w_2$. The branches of the tree are
$(u, v)$, $(v, w_1)$ and $(v, w_2)$. We assume we are given the
epigenome at each node in this tree, with the exception of the state
at position $n$ of node $v$. We are also given all paths for sites
other than site $n$. We want to make inferences about the state at
position $n$ of $v$, and also about the paths for edges incident on
$v(n)$. We require a method for sampling paths at site $n$ that share
a common state at their common end-point.

\citetext{hobolth2009simulation} reviewed and compared three
approaches to sample paths of discrete-state continuous-time Markov
chain conditional on end-point states, namely the rejection sampling,
direct sampling and uniformization. Method 1 in the previous
section is similar to uniformization, since the neighboring sites may
have state changes, and the rate of jumps depends on the contemporary
states of the neighboring sites. We proposed to use an averaged rate
to for the Poisson distribution, and let the acceptance probability to
do most of the work of correcting the proposal distribution towards the
posterior distribution. Method 2 is forward sampling, which is the
basis of rejection sampling.

\paragraph{Rejection sampling for the proposal:}
This approach first does forward sampling as described in Method 2
above for position $n$ along the branch $(u,v)$ and $(v, w_1)$, which
can be viewed as a single path that proceeds along two consecutive
edges. The sampled path provides a state $x$ corresponding to node
$v$.  Next, a path is sampled for edge $(v, w_2)$ assuming state $x$
at node $v$. If this path finishes with the known state at position
$n$ in $w_2$, we retain the 3 paths $L_{n}^{(u,v)}$, $L_{n}^{(v,w_1)}$
and $L_{n}^{(v,w_2)}$ as a proposal. The proposal is evaluated using
the same acceptance rule given above in equation
\eqref{eqn:rejection}.

%% Reject the path
%% unless the end state agree with the given state at node $a$ at
%% position $n$.  Then do forward sampling along the branch $(v, b)$,
%% with the accepted simulated state of node $v$. If the end state does
%% not agree with the given state at node $b$ at position $n$, start over
%% the sampling from the beginning, \textit{i.e.} start over from node
%% $u$. Because we are not taking site $n-2$ and $n+2$ into
%% consideration, we are not directly sampling from the posterior
%% distribution of the paths, therefore the rejection sampling procedure
%% is a way to propose viable paths, then we use the MCMC acceptance rule
%% \eqref{eqn:rejection} to reshape the proposal distribution to the
%% posterior distribution.

A couple of performance statistics should be measured for this
proposed sampling method. These include (1) the success rate for
generating valid paths that satisfy the end conditions; (2) the
acceptance rate of proposed valid paths in MCMC sampling.

\subsection{Posterior sampling of a path on the entire phylogenetic tree}
For a phylogenetic tree with known branch lengths and known
evolutionary context-dependent transition rates, how can we sample the
evolutionary path (on the entire tree) for one site from its posterior
distribution given observed evolutionary paths at all other sites and the
observed states at all leaf nodes for this site?

\paragraph{States at break points.}
The evolution process for this site is not homogeneous along the tree,
its transition rate matrix depends on the concurrent states of
neighboring sites. We can, however, partition each branch into
intervals, within each of which the states at the two neighboring
sites stay unchanged. Thus the evolutionary process for the site of
interest is homogeneous within each interval, and can be characterized
with a $2\times2$ transition rate matrix $Q_{bm}$ for the $m$-th
interval on branch $b$. The states of this site of interest at the
start and the end of a single interval, $s_{b,m-1}$ and $s_{b,m}$ are
connected through a transition probability matrix $P_{bm} =
\exp(Q_{bm}t_{bm})$.
The likelihood of observing leaf node states at this site
$\{s_n(v)\}$, given the evolutionary paths at other sites is
\[
\Pr(\{s_n(v)\} | L_{-n}) = \sum_{\{s_{bm}\}} \prod_{b} \prod_{m} P_{bm}(s_{b,m-1}, s_{b, m})
\]

\begin{figure}\label{fig:treebreak}
  \pgfimage{tree.pdf}\caption{Tree with break points as nodes}
\end{figure}

\paragraph{Bottom-up pruning algorithm.}
We can use Felsenstein's pruning algorithm to compute this
probability.  Let the phylogenetic tree be $\tau$ consisting of all
leaf nodes, internal nodes, and all break points between intervals on
individual branches as nodes.  Let $v$ be a node in the phylogenetic
tree, and $X_n(v)$ be the states of all leaf nodes that are
descendants of $v$ (Figure~\ref{fig:treebreak}) at site $n$ in the
genome. Let $u$ be the parent of $v$, and $w, z$ be the children of
$v$. Since the sequence contexts are known for the current site at all
intervals, the transition rate matrix for the state of the current
site is known and denoted with $Q_{\context{}(v)}$. Thus the
transition probability matrix between any parent-child pair $(u,v)$ is
known, which we denote with
\[
P_v = \exp(Q_{\context{}(v)}\ell_v).
\]
For node $v$, given that its parent $u$ has state $j$, let the
conditional probability of observing states $X_n(v)$ at
terminal descendants of node $v$ be $p_j(v)$, \textit{i.e.}
\[
p_j(v) = \Pr(X_n(v) | s_u=j).
\]
For notational convenience we define
\[
q_k(v) = \left\{
  \begin{array}{ll}
    \Pr(s_v = k) & \mbox{if $v$ is a leaf node,} \\
    \prod_{c\in \child{v}}  p_{k}(c) & \mbox{otherwise,} \\
  \end{array}\right.
\]
where $\Pr(s_v = k) \in \{0, 1\}$ when a state is observed, and
$\Pr(s_v = k) \in (0,1)$ when the observed data are continuous levels
representing a probability distribution over the state space. Then, we
can use Felsenstein's pruning algorithm to iteratively compute this probability
from bottom of the tree upwards to the root node:
\begin{equation}
  p_j(v) = \textstyle{\sum_{k}} \big(P_v(j,k) \times q_k(v)\big).
\end{equation}
In this way, we can compute the marginal posterior probability of observing the
states $X_n(v)$ at leaf nodes at site $n$, given the evolutionary
paths at all other sites:
\begin{equation}\label{eqn:leafmarginal}
\Pr(X_n(r) | L_{-n}) = \sum_j\Pr(s_r=j |L_{-n}) q_j(r)
\end{equation}
%%%
At the root node, the marginal posterior distribution of the state is
\begin{equation} \label{eqn:rootposterior}
\begin{aligned}
\Pr(s_r| L_{-n}, X_n(r)) = & \frac{\Pr(s_r, X_n(r)|L_{-n})}{\Pr(X_n(r)| L_{-n})} = \frac{\Pr(X_n(r)|s_r, L_{-n}) \Pr(s_r|L_{-n})}{\Pr(X_n(r)| L_{-n}) }  \\
= & \frac{ \big(\prod_{c\in \child{r}}  p_{s_r}(c)\big) \Pr(s_r|L_{-n})}{\Pr(X_n(r)| L_{-n}) },
\end{aligned}
\end{equation}
where $\Pr(s_r|L_{-n})$ is the initial distribution of the root node
state given its sequence context, which we can compute from the
stationary distribution of the root epigenomic sequence characterized
by either a Gibbs measure, or a Markov chain.

In summary, during the bottom-up Pruning algorithm \ref{alg:bottomup}, we compute
$p_j(v), q_k(v)$, and ultimately the marginal probabilities
 $\Pr(X_n(r) | L_{-n})$ and $\Pr(s_r|L_{-n}, X_n(r))$.

\begin{algorithm}[t]
  \begin{algorithmic}[1]
    \caption{Pruning algorithm with observed leaf states $X_n(r)$}\label{alg:bottomup}
    \FOR {node $v$ in post-order traversal}
    \IF {$v\neq r$}
    \IF {$v$ is a leaf node}
    \STATE Assign $q_k(v) \leftarrow I\{\Pr(s_v = k)\}$ for $k = 0, 1$.
    \ELSE
    \STATE Compute $q_k(v) \leftarrow \prod_{c\in \child{v}}  p_{k}(c)$
    \ENDIF
    \STATE Compute $p_j(v) \leftarrow \sum\limits_{k} \bigg(P_v[j,k] \times q_k(v)\bigg)$, for $j = 0, 1$.
    \ELSE
    \STATE Compute $\Pr(X_n(r) | L_{-n})$ as in
    \eqref{eqn:leafmarginal}, and $\Pr(s_r| L_{-n}, X_n(r))$ as in
    \eqref{eqn:rootposterior}
    \ENDIF
    \ENDFOR
  \end{algorithmic}
\end{algorithm}


\paragraph{Top-down posterior sampling at breakpoints.}
Let $S$ be the states at all internal nodes. The posterior probability
of all internal nodes taking the states $S$, given the leaf node
states $X_n(r)$ and the evolutionary paths at all other sites, is
\[
\Pr(S | X_n(r), L_{-n}) = \frac{\Pr(S, X_n(r) | L_{-n}) }{\Pr(X_n(r) | L_{-n})},
\]
where $\Pr(S, X_n(r) | L_{-n}) $ is the forward sampling probability
under the given transition probability matrices in individual
intervals determined by $L_{-n}$.
Therefore, we can do exact posterior sampling of the joint states at
the break points between the intervals on the phylogenetic tree through
direct sampling following Algorithm~\ref{alg:topdown}.

\begin{algorithm}[t]
  \begin{algorithmic}[1]
    \caption{Posterior sampling at breakpoints}\label{alg:topdown}
    \STATE Sample the root state from distribution $\Pr(s_r|L_{-n}, X_n(r))$ as in \eqref{eqn:rootposterior}.
    \FOR {internal node $v$ in pre-order traversal}
    \STATE Suppose its parent node $u$ has state $j$
    \STATE Sample state $k$ from the distribution
    \[k \sim \frac{1}{p_j(v)}\times P_v[j,k] \times q_k(v) \]
    \STATE Update the direct sampling probability with factor  $\frac{P_v[j,k]q_k(v)}{p_j(v)}$
    \ENDFOR
    \STATE The result is an sample $S$ of states at all internal states,
  which is sampled from their joint marginal posterior probability
  distribution $\Pr(S| X_n(r), L_{-n}) $.
  \end{algorithmic}
\end{algorithm}


%%%%%%
%%%%%%
\paragraph{Direct sampling of path between breakpoints with fixed states.}
Next, we sample the state transitions at this position within each
single time interval, during which its two neighbors have unchanged
states, and the initial and final states for this time interval are
fixed.  Focusing on one interval, let $0$ and $\tau$ be the end time
points.  Let an instance of the path within this interval be $L_n =
\{t_m\}_{m=0}^{M}$, which is a set of jumping times and the start and
end time points ($t_0 = 0$ and $t_M=\tau$). Let $\Delta_m = t_m -
t_{m-1}$ be the holding time between jumps. Let the homogeneous
transition rate matrix be $Q$ for this site in this time interval, and
$P(t)$ be the transition probability matrix for states at two time
points separated by time $t$. We use the direct sampling procedure
\cite{hobolth2009simulation} for endpoint-conditioned path sampling.

If the endpoints have the same state $i$, the probability that there
are no change during this time interval is
\begin{equation} \label{eqn:pnojump}
p_i = \frac{\exp(Q_{ii}\tau)}{P(\tau)_{ii}}.
\end{equation}
Then with probability $1-p_i$, at least one (actually two) state change occurs.

Let $W$ be the waiting time until the first jump.
\begin{equation}
\begin{aligned}
& \Pr(W < \tau | s(0) = i, s(\tau) = j)  \\
& =  \Pr(W < \tau, s(W) = \bar{i} | s(0) = i, s(\tau) = j)\\
& = \frac{\Pr(W < \tau, s(W) = \bar{i},s(\tau) = j | s(0) = i) } {\Pr(s(\tau) = j | s(0) = i)}  \\
& = \int_{0}^{\tau} Q_{i\bar{i}} \exp(Q_{ii}w) \frac{P(\tau-w)_{\bar{i}j}}{P(\tau)_{ij}} dw \\
& = \int_{0}^{\tau} f_{ij\tau}(w) dw, \text{~where~}
f_{ij\tau}(t) = Q_{i\bar{i}} \exp(Q_{ii}t) \frac{P(\tau-t)_{\bar{i}j}}{P(\tau)_{ij}}.
\end{aligned}
\end{equation}
The direct sampling procedure for a homogeneous continuous-time Markov path within an end-conditioned interval is detailed in Algorithm~\ref{alg:samplepath}.

\begin{algorithm}[t]
  \begin{algorithmic}[1]
    \caption{Direct sampling of end-conditioned path}\label{alg:samplepath}
    \STATE Suppose interval has length $\tau$, and start and end states $i, j$:
    \WHILE {$\tau > 0$}
    \IF {$i = j$}
    \STATE Sample $Z\sim \text{Bernoulli}(p_i)$, where
    $p_i$ is given in \eqref{eqn:pnojump}.
    \IF {$Z = 1$}
    \STATE Set $s(t) = i$ for all $0 < t < \tau$, and update $\tau \leftarrow 0$.
    \ENDIF
    \ENDIF
    \IF {$i \neq j$ or $Z = 0$}
    \STATE Sample the waiting time $W$ in state $i$ according to the continuous
    density $f_{ij\tau}(w)$, $0< w < \tau$, set $s(t) = i$ for $0 < t < W$
    \STATE Update $i \leftarrow \bar{i}$, $\tau \leftarrow \tau - W$.
    \ENDIF
    \ENDWHILE
  \end{algorithmic}
\end{algorithm}


\section{Discussion}

As noted earlier, the generalizations previously described for
phylo-HMMs have many similarities with the model we proposed. In
particular, these generalizations associate a binary variable with
each nucleotide in each sequence, including ancestral sequences.  The
most notable difference is our emphasis on interpretability of the
horizontal relationships, specifically that an individual epigenome
(extant or ancestral) follows a Markov process. Another difference is
our adoption of a homogeneous continuous time process, which is both
built into the model, and exploited by our inference
procedure. Phylogenetic HMMs are applied to determine states along
aligned genomes where those genome sequence evolve according to a
particular substitution model. Our notion of the epigenome corresponds
more closely to the idea of ``conserved'' and ``non-conserved''
states, but the details of our approaches are more similar to the
substitution process of nucleotides, rather than general
time-dependent graphical models.

\clearpage

\bibliographystyle{namedplus}
\bibliography{biblio}

\clearpage

\section{Gradients of the log-likelihood}

We treat $\{\log\lambda_c: c = 0,1,2,3,5\}$ as free parameters as they
can take any real value.
\begin{equation}
  \begin{aligned}
    \frac{\partial l}{\partial \log\lambda_0} &= J_0 - D_0\lambda_0 + J_7 - D_7\lambda_7\\
    \frac{\partial l}{\partial \log\lambda_2} &= J_2 - D_2\lambda_2 - J_7 + D_7\lambda_7 \\
    \frac{\partial l}{\partial \log\lambda_5} &= J_5 - D_5\lambda_5 + J_7 - D_7\lambda_7\\
    \frac{\partial l}{\partial \log\lambda_1} &= J_1 + J_4 - (D_1 + D_4)\lambda_1 - 2J_7 + 2D_7\lambda_7\\
    \frac{\partial l}{\partial \log\lambda_3} &= J_3 + J_6 - (D_3 + D_6)\lambda_3 + 2J_7 - 2D_7\lambda_7
  \end{aligned}
\end{equation}

\appendix


\end{document}

%% \begin{enumerate}
%% \item Let $t \leftarrow 0$, and initialize all paths $L_n = \{x_n(0),
%%   k_n=0, T_n=\emptyset, t\}$, for $n=1,\ldots, N.$
%% \item While $t < T$:
%%   \begin{enumerate}
%%   \item Generate $y\sim \text{Exp}(-M_{x(t)x(t)})$, where
%%     $-M_{x(t)x(t)} = \sum_{i,j,k}c_{ijk}(x(t))\lambda_{ijk}$.
%%     \begin{itemize}
%%     \item[If] $t+x < T$ then:
%%       \begin{itemize}
%%       \item Choose triple $ijk \in \{0,1\}^3$ with
%%         probability proportional to $c_{ijk}(x(t))\lambda_{ijk}$.
%%       \item Uniformly sample a position $n$ among the $c_{ijk}(x(t))$ positions having pattern $ijk$.
%%       \item $x(t+y) \leftarrow x(t)[1..n-1]\overline{x(t)_n}x(t)[n+1..N]$.
%%       \item Add jump time to the path of position $n$:
%%         \[
%%         k_n \leftarrow k_n + 1; ~ T_n \leftarrow T_n\cup \{t+x\}.
%%         \]
%%       \end{itemize}
%%     \item[Else:] $x(T) \leftarrow x(t)$.
%%     \end{itemize}
%%   \item $t \leftarrow t+y$
%%   \end{enumerate}
%% \end{enumerate}


% \section{Simulation scheme 1}
% We are going to simulate a full history of epigenome evolution for a
% time interval $[0, t]$.

% \begin{enumerate}
% \item Simulate the starting methylome using a binary-state Markov model
% \item Initialize all paths $L_n = \{s_n(0), k_n=0, T_n=\emptyset, t\}$, for $n=1,\ldots, N.$
% \item (For simplicity, fix the paths $L_1$ and $L_N$ as initialized.) For
%   site $n = 2, \ldots, N-1$, simulate $L_n$ given the current paths of
%   $L_{n-1}$ and $L_{n+1}$:
%   \begin{itemize}
%   \item Collect the time intervals from site $n-1$ and site $n+1$, so
%     that within each of the intervals the states of the neighboring
%     sites are unchanged. Let the intervals be represented by a sorted
%     array $\{t_0, t_1, \cdots, t_M\}$.
%   \item For $m = 1, \ldots, M$:
%     \begin{itemize}
%     \item Let $X$ be a random variable from an exponential
%       distribution, representing the jumping time of the middle site given
%       that the states of its two neighbors are constant.  For a time
%       interval $[t_{m-1}, t_m]$ during which the neighboring sites' states
%       are unchanged, suppose $X\sim \text{Exp}(\lambda)$. The parameter
%       $\lambda$ is a function of current state and the states of the two
%       neighboring sites.
%     \item Let $t = t_{m-1}$, and a sample value of $X=x$.
%     \item While $t +x < t_m$:
%       \begin{enumerate}
%       \item[(1)] Add 1 to the number of jumps $k_n \leftarrow k_n +1$.
%       \item[(2)] Update $T_n \leftarrow T_n\cup\{t+x\}$.
%       \item[(3)] Update $t \leftarrow  t+x$,
%       \item[(4)] Sample another value of $X=x$ from $\text{Exp}(\lambda)$.
%       \end{enumerate}
%     \end{itemize}
%   \end{itemize}
% \item Repeat step 3, until the epigenome summary statistics converge to
%   a stable distribution.
% \end{enumerate}

%% \subsubsection{Scaling parameters}

%% After the estimates are made, we can scale the transition rates and
%% branch lengths to have unit branch length corresponding to 1 expected
%% transition per site.  Given $\lambda_{c}$ according to
%% \eqref{eqn:rel}, the ratios $\frac{\lambda_{010}}{\lambda_{000}}$ and
%% $\frac{\lambda_{001}}{\lambda_{011}}$ uniquely determine the
%% stationary distribution for the epigenome described by a Gibbs measure
%% of the form \eqref{eqn:stationary}. The Gibbs measure for the
%% epigenome sequence, in turn, is equivalent to the Markov chain
%% \eqref{eqn:gibbs2markov}. Given the Markov chain formulation, we can
%% compute the expected frequency of triplet patterns
%% \[
%% p_{ijk} = \pi_i \times T_{ij} \times T_{jk}.
%% \]
%% Then the expected number of changes per position per
%% unit time is $\mu = \sum_{ijk}p_{ijk}\lambda_{ijk}$. We can scale transition
%% rates and branch lengths as follows:
%% \begin{equation}\label{eqn:tidentifiable}
%%   \begin{aligned}
%%     \lambda_{ijk} &\leftarrow \lambda_{ijk}\times \mu \\
%%     t_b &\leftarrow \sum_{m=1}^{M_b} (t_{b,m} - t_{b,m-1}) \times \frac{1}{\mu}.
%%   \end{aligned}
%% \end{equation}
%% %%


%% \paragraph{Method 1:} First, sample a starting state $s_0$ from a
%% Bernoulli distribution with probabilities $(\pi_0, \pi_1)$. Then,
%% propose a number $K$ of jumps from a Poisson distribution with rate
%% parameter $\lambda = \sum\lambda_{ijk}/8$, which is chosen to
%% approximate average mutation rate among different contexts. Given $K$,
%% sample jump times uniformly on the time interval $(0, t)$; In other
%% words, from the Dirichlet distribution with concentration parameters
%% all equal to 1. Therefore, the probability (density) of proposing a
%% specific path $L' = \{s_0, K, \{t_k\}_{k=1}^K, t\}$ is
%% \[
%% q(L') = \pi_{s_0} \frac{\lambda^K \exp(-\lambda)}{K!} \frac{1}{(K-1)!}.
%% \]
%% This is known as an \textit{independence sampler}\cite{}. If the
%% current path at position $n$ is $L_n$, then we accept the proposed
%% path $L'$ with probability
%% \begin{equation}\label{eqn:rejection}
%%   \alpha(L') = \min\left\{\frac{\pi(L', L_{-n})/q(L')}{\pi(L_n, L_{-n})/q(L_n)}, 1\right\},
%% \end{equation}
%% where $\pi$ is the complete data likelihood function \eqref{eqn:lik}.

%% \paragraph{Method 2:} The method outlined above uses a proposal
%% distribution that is approximately uniform. This can be highly
%% inefficient, {\it i.e.} lead to a low rate of acceptance, although
%% \citetext{jensen2000probabilistic} used a sampling procedure in the
%% same spirit. If we use more information from the paths at neighboring
%% sites, we may be able to improve sampling efficiency. The neighboring
%% paths $L_{n-1}$ and $L_{n+1}$ partition the evolutionary time interval
%% $(0,t)$ into time segments during which the states at positions $n-1$
%% and $n+1$ do not change.

%% \begin{itemize}
%% \item Collect the time intervals from site $n-1$ and site $n+1$, so
%%   that within each of the intervals the states of the neighboring
%%   sites are unchanged. Let the intervals be represented by a sorted
%%   array $\{t_0, t_1, \cdots, t_M\}$.
%% \item Initialize proposal probability $p\leftarrow 1$.
%% \item For $m = 1, \ldots, M$:
%%   \begin{itemize}
%%   \item Let $X$ be a random variable from an exponential distribution,
%%     representing the jumping time of the middle site given that the states
%%     of its two neighbors are constant.  For a time interval $[t_{m-1},
%%     t_m]$ during which the neighboring sites' states are unchanged,
%%     suppose $X \sim \mathit{Exp}(\lambda_{ijk})$, where $i$ and $k$ are the
%%     states of the two neighboring sites in this time interval, and $j$ is
%%     the starting state of position $n$. Let $f_{\lambda}$ be the p.d.f. of
%%     Exponential distribution $\mathit{Exp}(\lambda)$.
%%   \item Let $t = t_{m-1}$, and a sample value of $X=x$.
%%   \item While $t + x < t_m$:
%%     \begin{enumerate}[label={(\arabic*)}]
%%     \item Update proposal probability $p \leftarrow p\times f_{\lambda_{ijk}}(x)$.
%%     \item Update the state of the center site $j \leftarrow \bar{j}$.
%%     \item Add 1 to the number of jumps $k_n \leftarrow k_n +1$.
%%     \item Update $T_n \leftarrow T_n\cup\{t+x\}$.
%%     \item Update $t \leftarrow  t+x$,
%%     \item Sample another value of $X=x$ from $\mathit{Exp}(\lambda_{ijk})$.
%%     \end{enumerate}
%%   \item Update proposal probability $p \leftarrow p\times \Pr_{\lambda_{ijk}}(X > t_m - t )$.
%%   \end{itemize}
%% \end{itemize}

%% The proposal probability density function $q()$ is thus the product of
%% the appropriate Exponential distribution probability densities for the
%% holding times at position $n$, and is calculated as $p$. The proposal
%% distribution is independent of any current guess of the
%% path. Therefore, this is also an \textit{independence sampler}. The
%% rejection rule stays the same, as in (\ref{eqn:rejection}).

%% \paragraph{Note:} For two versions of complete evolutionary history
%% that only differ at one position, their difference in likelihood is
%% determined by the paths at that position, and two positions to each
%% side, \textit{i.e.} 5 positions in total. Because the total time spent
%% in each triplet context \eqref{eqn:Dijk} is altered at the center
%% position and its two neighboring positions. Let $n$  be the center position.
%% Let $D_{n,c}$ be the total time spent in triplet context $c$ at position $n$
%% in the current instance of evolutionary paths, and let $D'_{n,c}$ be that of
%% the new proposed evolutionary path. Similarly let $J_{n,c}$ and $J'_{n,c}$ be
%% the total number of jumps out of context $c$ at position $n$ in the
%% current and proposed paths. Thus, the ratio of complete data
%% likelihood in acceptance probability \eqref{eqn:rejection} can be
%% computed as
%% \[
%% \frac{\pi(L', L_{-n})}{\pi(L_n, L_{-n})} = \prod_{c}\prod_{s=n-1}^{n+1}
%% \lambda_c^{J'_{s,c} - J_{s,c}} \exp(-(D'_{s,c}-D_{s,c})\lambda_c).
%% \]

%% %% JQU: current goal of coding is to make sure after one pass of updates,
%% %% the parameters can be recovered from the updated paths.

%% % \paragraph{Note:} A jump at time $\tau$ within the path $L_{n}$
%% % will affect the probability of all jumps that occur after time $\tau$,
%% % regardless of their position within the epigenome. The is because such
%% % a jump affects $\{c_{ijk}(t)\}$ for any $t > \tau$. Therefore, we
%% % cannot cancel out factors in the likelihood function. However, we may
%% % try to make things easy by only considering the evolutionary paths at
%% % positions $n-2$, $n-1$, $n+1$, and $n+2$.
