\documentclass[11pt]{article}

\usepackage{fullpage,times,namedplus,pgf}
\usepackage{amsmath,amssymb}
\usepackage{kbordermatrix}% http://www.hss.caltech.edu/~kcb/TeX/kbordermatrix.sty
\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator{\E}{\mathbb{E}}

\newcommand{\myroot}{\ensuremath{\mathrm{root}}}

\title{A model of epigenome evolution}

\begin{document}
\maketitle

\begin{abstract}
  Epigenetic marks along the mammalian genome are organized into
  alternating genomic domains bearing and lacking the mark. The
  location and size of domains enriched for an epigenetic mark are
  indicative of the presence, function and activity of regulatory
  elements and the chromatin states. Comparative epigenomic studies
  aim to resolve the evolutionary history of regulatory elements by
  comparing epigenomic profiles in multiple species.  However,
  computational methods for comparing epigenetic marks at high
  resolution, inferring evolution rates along different phylogenetic
  lineages and reconstructing the evolutionary history are still
  limited.  In this study, we aim to establish a simulation, sampling
  and inference framework for studying the evolution of the genomic
  distribution of an epigenetic from the profiles of multiple extant
  species.  We model the profile of an epigenetic mark in a species
  with a two-state Markov chain, and model the evolution of an
  epigenomic sequence with a continuous-time Markov chain, where
  instantaneous transition rates at a site is dependent on the
  contemporary states of its neighboring sites. We use a MCMC sampling
  method for estimating the context-dependent transition rates and
  inferring the evolutionary history that lead to diverse profiles in
  extant species from a common ancestral epigenome.  We show with
  applications to DNA methylation and histone modification profiles
  that our methods can reveal both genome-wide evolutionary features
  through estimates of the model parameters and high-resolution
  evolutionary patterns in local regions through posterior sampling of
  the evolutionary history.
%%   The epigenome of vertebrate cells encodes regulatory complexity
%%   equivalent to that observed in temporal and spatial patterns of
%%   cellular phenotypes. One highly useful simplification has been to
%%   consider the epigenome as indicating active and inactive regulatory
%%   regions, including promoters and enhancers. For a fixed cell type
%%   shared across species, one may ask whether the epigenomic state of
%%   orthologous sequences has changed as species evolve.
%% %%%% Mention previous work like phyloHMM
%%   Although we know little about the forces shaping regulatory
%%   evolution, it seems as though regulatory functions turns over much
%%   more rapidly than genes, and the number of states required to model
%%   this as a hidden Markov model would be exponential in the number of
%%   species.
%% %%%% Say what we present
%%   Here we present a model of epigenomic evolution that incorporates
%%   interdependence of neighboring states so that the stationary
%%   distribution is a Markov process within a species. Our model
%%   includes parameters that determine the rate of birth and death of
%%   active regions, along with their rate of expansion and contraction.
\end{abstract}

\section{Introduction}

Assume the epigenome is a sequence of auto-correlated binary-state
random variables. The auto-correlation reflects the organization of
the epigenome as alternating domains bearing a certain modification or
not. Let the epigenomic sequence be $S=s_1s_2\ldots s_N$. As a
graphical model, neighboring sites are connected with undirected
edges. $S$ evolves over time. Assume the stationary distribution for
the epigenome has a Gibbs distribution that factorize over pairs of
neighboring sites:
\begin{equation}\label{eqn:stationary}
\Pr(S) = \frac{1}{Z} \exp\big\{\phi(s_1) +\sum_{n=1}^{N-1}\phi(s_n, s_{n+1}) + \phi(s_N)\big\}
\end{equation}

The evolution of an individual site is context-dependent. The
instantaneous mutation rate from state $s$ to the alternative state
$\bar{s}$ is $\gamma(l, s, r)$, where $l$, $s$ and $r$ are the states
of three consecutive sites.  For a time interval $[0,t)$, given that
the states at positions $l$ and $r$ do not change, then the state
at site $s$ follows a continouse-time Markov process, and the holding
time thus follows an exponential distribution. An observation of
this path can be summarized as
\[
  L = \big\{s(0), k, \{t_i\}_{i=1}^{k}, t \big\},
\]
where $s(0)$ is the state at time 0, $k$ is the total number of jumps,
$t_i$ is the time of occurrence for the $i$-th jump, and $t$ is the total
length of the time interval.

Suppose $L_l$ and $L_r$ are paths for distinct sites. The union of their jump times
\[
  \{0, t\} \cup \{t_{li}\}_{i=1}^{k_l} \cup \{t_{ri}\}_{i=1}^{k_r}
\]
gives a set of time points such that when ordered, any pair
of consecutive times in this union defines an interval during
which the joint state at sites $l$ and $r$ remains
constant.

\section{The model}

\subsection{Stationary Gibbs measure as Markov chain}

The stationary distribution in \eqref{eqn:stationary} is equivalent to
the distribution of a Markov chain. We can derive the relationship
between the factors in \eqref{eqn:stationary} and the transition
probabilities of a Markov chain. The pair-wise potentials are
$Q(a,b)=\exp(\phi(a, b))$, where $a, b\in\{0,1\}$ are binary
states. The largest eigenvalue of $Q$ is
\[
q=\frac{1}{2}\left(Q_{00}+Q_{11} +\sqrt{\Delta}\right), \text{ where }
\Delta=(Q_{00} - Q_{11})^2 + 4Q_{01}Q_{10}.
\]
Let $h$ be a right eigenvector of $Q$ corresponding to $q$, then we have
\[
\frac{h_0}{h_1} = \frac{Q_{01}}{q-Q_{00}} = \frac{q-Q_{11}}{Q_{10}}
\]
Then for states $a, b\in \{0, 1\}$, the Markov chain transition matrix
is
\[
  T(a, b) = \frac{Q_{ab}h_b}{qh_{a}}.
\]
More specifically,
\begin{equation} \label{eqn:gibbs2markov}
  \begin{array}{ll}
    T(1,1) = \displaystyle\frac{2Q_{11}}{Q_{00}+Q_{11}+\sqrt{\Delta}}, &
    T(0,0) = \displaystyle\frac{2Q_{00}}{Q_{00}+Q_{11}+\sqrt{\Delta}}, \\[2em]
    T(0,1) = \displaystyle\frac{4Q_{01}Q_{10}}{(Q_{00}+\sqrt{\Delta})^2 -Q_{11}^2}, &
    T(1,0) = \displaystyle\frac{4Q_{01}Q_{10}}{(Q_{11}+\sqrt{\Delta})^2 -Q_{00}^2}.
  \end{array}
\end{equation}
The expected fraction of an epigenome residing within
functional domains is thus:
\[
1- \frac{2Q_{01}Q_{10}}{(Q_{00}-Q_{11})^2 + 4Q_{01}Q_{10} +
  (Q_{11}-Q_{00})\sqrt{\Delta}}.
\]

\subsection{Relating the substitution model and the stationary distribution}

\citetext{jensen2000probabilistic} gave a sufficient condition for the
continuous time evolutionary model $\gamma$ to have a stationary
distribution determined by $\phi$. In particular,
%%% Prop 1 of J-P 2000
\begin{equation}\label{eqn:prop1}
  \frac{\gamma(l, s, r)}{\gamma(l, \bar{s}, r)} =
  \frac{\exp(\phi(l, \bar{s})+ \phi(\bar{s}, r))}{\exp(\phi(l, s)+ \phi(s, r))},
\end{equation}
%%% ADS: not sure this statement below is correct
which can be derived from the reversibility property of the stationary
distribution.

Proposition 2 of \citetext{jensen2000probabilistic} provides a way of
specifying $\gamma$ from $\phi$: substitution rates $\gamma$ satisfy
the relation \eqref{eqn:prop1} if and only if they can be written in
the form
\begin{equation}\label{eqn:prop1}
 \log(\gamma(l, s, r)) = -\psi(l, s, r) + \ell(l, r).
\end{equation}
This criteria was introduced in the context of an arbitrary number of
states, and the $\ell$ function can be understood as $\ell(s, t; l,
r)$, for two states $s$ and $t$, which is symmetric in $(s, t)$. In
our setting of modeling epigenomic features, the states are binary and
$\ell$ is only a function of the two neighboring states $(l,r)$.
Moreover, we can directly verify that if we define substitution rates as
\begin{equation}
\log (\gamma(l, s, r)) =  \ell(l, r) + (\phi(l, \bar{s}) +\phi(\bar{s}, r)),
\end{equation}
then they satisfy the condition of \eqref{eqn:prop1}.

\subsection{Parameterization and interpretation}

To help interpret this model we can organzie the substitution rates in
an $8\times8$ matrix:
\begin{equation}\label{eqn:Gamma}
\renewcommand{\kbldelim}{(}% Left delimiter
\renewcommand{\kbrdelim}{)}% Right delimiter
  \Gamma = \kbordermatrix{
        & 000 & 010 & 001 & 011 & 100 & 110 & 101 & 111 \\
    000 & \cdot & a & 0 & 0 & 0 & 0 & 0 & 0 \\
    010 & b & \cdot & 0 & 0 & 0 & 0 & 0 & 0 \\
    001 & 0 & 0 & \cdot & c & 0 & 0 & 0 & 0 \\
    011 & 0 & 0 & d & \cdot & 0 & 0 & 0 & 0 \\
    100 & 0 & 0 & 0 & 0 & \cdot & c & 0 & 0 \\
    110 & 0 & 0 & 0 & 0 & d & \cdot & 0 & 0 \\
    101 & 0 & 0 & 0 & 0 & 0 & 0 & \cdot & e \\
    111 & 0 & 0 & 0 & 0 & 0 & 0 & f & \cdot
  }
\end{equation}
We may interpret the values in $\Gamma$ in the following way. The
values $a$ and $b$ correspond to the ``birth'' and ``death'' of an
epigenomic feature during evolution of the epigenome, respectively.
The value $e$ corresponds to the merging of two features into a single
contiguous interval ($101\rightarrow 111$). Conversely, the value $f$
corresponds to epigenomic features ``splitting'' and becoming two
separate intervals ($111\rightarrow 101$). The remaining values, $c$
and $d$, correspond to existing intervals either becoming wider or
more narrow (expanding or contracting), and both of these parameters
appear twice because we do not distinguish beetween
expanding/contracting to the left from the right. This left-right
symmetry is biologically reasonable, particularly for epigenomic
features that have no directionality. When epigenomic features are near
particular genome annotations, like transcription start sites, one
might argue that directionality should be considered.

If an epigenome evolves according to substitution rates $\Gamma$ with
stationary distribution \eqref{eqn:stationary}, then the condition in
\eqref{eqn:prop1} holds, and we have the following constraints for the
rates in the above matrix:
\begin{equation}\label{eqn:constraint}
  ad^2e=bc^2f.
\end{equation}
So given substitution rates $a,b,c,d$, we have the following
relationships between horizontal potentials:
\begin{equation}\label{eqn:rel}
  \phi(0,0) = \phi(0,1) +\frac{1}{2}\log\left(\frac{b}{a}\right), ~\text{ and }~
  \phi(1,1) = \phi(0,1) +\frac{1}{2}\log\left(\frac{bc^2}{ad^2}\right).
\end{equation}

In summary, the substitution rate matrix can have 5 free
parameters. If we add a constraint on the expected number of changes
per unit time, then the model will have only 4 free parameters. The
two ratios $b/a$ and $bc^2/(ad^2)$ then uniquely
determine the stationary distribution \eqref{eqn:stationary} through
equation \eqref{eqn:rel}.

% \section{Simulation scheme 1}
% We are going to simulate a full history of epigenome evolution for a
% time interval $[0, t]$.

% \begin{enumerate}
% \item Simulate the starting methylome using a binary-state Markov model
% \item Initialize all paths $L_n = \{s_n(0), k_n=0, T_n=\emptyset, t\}$, for $n=1,\ldots, N.$
% \item (For simplicity, fix the paths $L_1$ and $L_N$ as initialized.) For
%   site $n = 2, \ldots, N-1$, simulate $L_n$ given the current paths of
%   $L_{n-1}$ and $L_{n+1}$:
%   \begin{itemize}
%   \item Collect the time intervals from site $n-1$ and site $n+1$, so
%     that within each of the intervals the states of the neighboring
%     sites are unchanged. Let the intervals be represented by a sorted
%     array $\{t_0, t_1, \cdots, t_M\}$.
%   \item For $m = 1, \ldots, M$:
%     \begin{itemize}
%       \item Let $X$ be a random variable from an exponential
%         distribution, representing the jumping time of the middle site given
%         that the states of its two neighbors are constant.  For a time
%         interval $[t_{m-1}, t_m]$ during which the neighboring sites' states
%         are unchanged, suppose $X\sim \text{Exp}(\lambda)$. The parameter
%         $\lambda$ is a function of current state and the states of the two
%         neighboring sites.
%       \item Let $t = t_{m-1}$, and a sample value of $X=x$.
%       \item While $t +x < t_m$:
%         \begin{enumerate}
%         \item[(1)] Add 1 to the number of jumps $k_n \leftarrow k_n +1$.
%         \item[(2)] Update $T_n \leftarrow T_n\cup\{t+x\}$.
%         \item[(3)] Update $t \leftarrow  t+x$,
%         \item[(4)] Sample another value of $X=x$ from $\text{Exp}(\lambda)$.
%         \end{enumerate}
%       \end{itemize}
%   \end{itemize}
% \item Repeat step 3, until the epigenome summary statistics converge to
%   a stable distribution.
% \end{enumerate}


\section{Simulation scheme}

We model the evolution of the entire sequence that with a continuous
time Makov chain that allows instantaneous jump from one sequence to
another only if they differ at one position. The jumping rate at each
position is dependent on its state and the states of its neighboring
sites. We assume that the states of the first and last sites are
fixed throughout evolution.

Consider the $2^N \times 2^N$ transition rate matrix $M$, for any
methylome $a$ and methylome $b$ that have a single difference at a
position where $a$ has state $j$ and $b$ has state $\bar{j}$, and the
neighboring positions have states $i$ and $k$ in both methylomes, the
rate of such a jump is $\lambda_{ijk} = \gamma(i, j, k) = \Gamma_{ijk,
i\bar{j}k}$.

Given the current methylome $a$, the holding time
\[
  X_a\sim Exp(-M_{aa})
\]
is an exponential variable. Its rate parameter $-M_{aa}$ is the
sum of all instantaneous rates for jumps from $a$ to a methylome that
only differs with $a$ at one position:
\[
  -M_{aa} =  \sum\limits_{i,j,k}c_{ijk}(a)\lambda_{ijk},
\]
where $c_{ijk}(a) = \sum_{n=1}^{N-2}I(a_{n}=i, a_{n+1}=j, a_{n+2}=k)$
is the total number of the tripplet pattern $ijk$ in methylome $a$.

Given that the first jump happened, the probability that the jump
occurred in the context of $ijk$ is proportional to
$c_{ijk}(a)\lambda_{ijk}$. Given that a jump happend in context $ijk$,
the jump is equally likely among positions with this context.
%%
The expected number of changes per site per unit time is
$\sum\limits_{ijk}\pi_{ijk}\lambda_{ijk}$, where $\pi_{ijk}$ is the
stationary probability of pattern $ijk$ in the methylome.

In summary, we have the following simulation procedure for the evolution
process for a methylome with $N$ sites over time interval $[0, T]$, given
that the initial methylome is $a(0)$:
\begin{enumerate}
\item Let $t \leftarrow 0$, and initialize all paths $L_n = \{a_n(0),
  k_n=0, T_n=\emptyset, t\}$, for $n=1,\ldots, N.$
\item While $t < T$:
  \begin{enumerate}
  \item Generate $x\sim \text{Exp}(-M_{a(t)a(t)})$, where
    $-M_{a(t)a(t)} = \sum\limits_{i,j,k}c_{ijk}(a(t))\lambda_{ijk}$.
    \begin{itemize}
    \item[If] $t+x < T$:
      \begin{itemize}
      \item Choose pattern $ijk$ from $\{ijk: i,j,k\in\{0,1\}\}$ with
        probability proportional to $c_{ijk}(a(t))\lambda_{ijk}$.
      \item Scan methylome $a(t)$, uniformaly choose one position $n$
        out of the $c_{ijk}$ positions all with the pattern $ijk$ in $a(t)$.
      \item Set $a(t+x) \leftarrow a(t)_{1\ldots n-1} \overline{a(t)_n}
        a(t)_{n+1 \ldots N}$.
      \item Add jump time to the path of position $n$:
        $$k_n \leftarrow k_n +1, ~~ T_n \leftarrow T_n\cup \{t+x\}.$$
      \end{itemize}
    \item[Else] $a(T) \leftarrow a(t)$.
    \end{itemize}
  \item $t \leftarrow t+x$
  \end{enumerate}
\end{enumerate}

\section{Parameter inference}

When we're given the complete epigenome evolution path from time $0$
to time $t$, can we effectively recover the initial distribution and
mutation parameters and evolutionary time? We are interested in the
parameters describing the evolutionary process, which are the mutation
rates $\{\lambda_{ijk}\}$. These parameters will be inferred from the
state changes in the evolutionary path of the entire
epigenome. Meanwhile, we do not require the process to be stationary,
so we are also interested in the epigenome properties at time 0, which
are characterized by the Markov chain transition probabilities
$T_{0}$. These transition probabilities are easy to infer given the
complete observations at the time-0 epigenome.

Let $c_{ij} = \sum_{n=1}^{N-1}I\{s_n(0) =i, s_{n+1}(0)=j\}$. Then
\[ \hat{T}(0, 0) = \frac{c_{00}}{\sum_{n=1}^{N-1}I\{s_n(0) = 0\}}, ~
\hat{T}(1,1) = \frac{c_{11}}{\sum_{n=1}^{N-1}I\{s_n(0) = 1\}}.
\]

Let's first assume that the time span of this complete evolutionary
history is known, \textit{i.e.} the value of $t$ is give.

Recall that $L_n = \{s_n(0), K, \{t_k\}_{k=1}^K, t\}$ is a full path
at position $n$ in the epigenome. We can pool all the jumping times at
all positions in to a sorted sequence $J = \{(t_m, \text{pos}_m,
\text{context}_m) \}_{m=1}^{M}$, where $\text{pos}_m$ is the position of the
$m$-th jump in the entire evolutionary history of the epigenome,
$\text{context}_m$ is the 3-tuple context of the mutation.

Let $\Delta_m = t_m - t_{m-1}$ be the holding time before the $m$-th
jump. Then $\Delta_m$ is an exponential variable
\[
\Delta_m \sim \text{Exp}(\lambda_m), \text{where } \lambda_m = \sum\limits_{i,j,k}c_{ijk}(t_m - \epsilon)\lambda_{ijk},
\]
where constant
$\epsilon \in (0, \min\limits_{1\le m \le M}\{\Delta_{m}\})$ so that
$c_{ijk}(t_m - \epsilon)$ is the sequence
context distribution between the $(m-1)$th jump and the $m$-th jump.

The likelihood function for parameters $\{\lambda_{ijk}\}$ is thus
\begin{equation}\label{eqn:lik}
\begin{aligned}
L &= \prod\limits_{m=1}^{M} \lambda_m \exp(-\lambda_m\Delta_m) \times \frac{\lambda_{\text{context}_m}}{\lambda_m} \\
&=\prod\limits_{m=1}^{M}\lambda_{\text{context}_m}\exp(-\lambda_m\Delta_m)
\end{aligned}
\end{equation}

The log-likelihood function is
\begin{equation}\label{eqn:loglik1}
\begin{aligned}
l & = \sum\limits_{ijk} \big(
\sum_{m=1}^M{\log\lambda_{ijk}\times I_{\{ \text{context}_m = ijk\}} - c_{ijk}(t_m-\epsilon)\lambda_{ijk}\Delta_m } \big) \\
& = \sum\limits_{ijk} \big(J_{ijk}\log\lambda_{ijk} - D_{ijk}\lambda_{ijk} \big)
\end{aligned}
\end{equation}
where $J_{ijk} = \sum_{m=1}^M I_{\{\text{context}_m = ijk\}}$, and $D_{ijk} = \sum_{m=1}^Mc_{ijk}(t_m-\epsilon)\Delta_m$.

Constraints on the mutation rates $\{\lambda_{ijk}\}$ as indicated in
equations (\ref{eqn:Gamma},\ref{eqn:constraint}) are
\begin{equation}\label{eqn:constraints}
\left\{
  \begin{array}{ll}
    \lambda_{001} = \lambda_{100}\\
    \lambda_{011} = \lambda_{110}\\
    \lambda_{000}\lambda_{110}^2\lambda_{101} = \lambda_{010}\lambda_{100}^2\lambda_{111}
  \end{array}
\right.
\end{equation}

Then the log-likelihood function (\ref{eqn:loglik1}) becomes
\begin{equation}\label{eqn:loglik2}
\begin{aligned}
l = & J_{000}\log\lambda_{000} - D_{000}\lambda_{000} + \\
& J_{010}\log\lambda_{010} - D_{010}\lambda_{010} + \\
& J_{101}\log\lambda_{101} - D_{101}\lambda_{101} + \\
& (J_{100} + J_{001})\log\lambda_{001} - (D_{100}+D_{001})\lambda_{001} +  \\
& (J_{011} + J_{110})\log\lambda_{011} - (D_{011}+D_{110})\lambda_{011} +  \\
& J_{111}\log(\frac{\lambda_{000}\lambda_{011}^2\lambda_{101}}{\lambda_{010}\lambda_{001}^2}) - D_{111}\frac{\lambda_{000}\lambda_{011}^2\lambda_{101}}{\lambda_{010}\lambda_{001}^2}.
\end{aligned}
\end{equation}

When the time span of this complete evolutionary history is unknown,
the value of $t$ is also a model parameter to be estimated. We assume
that all the jumping times are expressed as a fraction of $t$. Then
the problem of estimating the mutation rates and evolutionary time
becomes identifiable. We need an extra constraint -- the unit branch
length corresponds to 1 expected mutation per site. This is a common
constraint in phylogenetic studies.

Here we explain how to formulate this constraint on the
$\{\lambda_{ijk}\}$ parameters. Given $\{\lambda_{ijk}\}$, according
to (\ref{eqn:rel}) $\frac{\lambda_{010}}{\lambda_{000}}$ and
$\frac{\lambda_{001}}{\lambda_{011}}$ can uniquely determine the
stationary distribution of the epigenome that is described with a
Gibbs measure of form (\ref{eqn:stationary}).  The Gibbs measure for
the epigenomic sequence, in turn, is equivalent to a Markov chain
(\ref{eqn:gibbs2markov}). Given the Markov chain formulation, we can
easily compute the expected abundance of triplet patterns $p_{ijk} =
\frac{1}{N}\E(c_{ijk})$, where $c_{ijk}$ is the frequency of the
triplet pattern in an epigenomic sequence of length $N$ sampled from
the Gibbs distribution. Then, we can compute the expected number of
changes per position per unit time as
$\sum_{ijk}p_{ijk}\lambda_{ijk}$. When the evolutionary time is also
unknown, we add the following constraint:
\begin{equation}\label{eqn:tidentifiable}
\sum_{ijk}p_{ijk}\lambda_{ijk} = 1,
\end{equation}
where $\{p_{ijk}\}$ as explained above are functions of
$\{\lambda_{ijk}\}$.  We maximize the log-likelihood
(\ref{eqn:loglik1}) over $\{\lambda_{ijk}\}\cup\{t\}$ under the
constraints (\ref{eqn:constraints}) and (\ref{eqn:tidentifiable}),
where the unknown parameter $t$ is buried within $\{D_{ijk}\}$ in
(\ref{eqn:loglik1}).


\paragraph{Posterior distribution of a path given two neighboring paths} We
assume that the model parameters $\{\lambda_{ijk}\}$ and total
evolutionary time $t$ are known, the starting state of the epigenome
are known except for position $n$, and that the jumping times $J =
\{(t_m, \text{pos}_m, \text{context}_m) \}_{m=1}^{M}$ are known for
all positions except position $n$. How can we estimate the posterior
distribution of path $L_{n}$ (a sequence of jumping times)?
\[
\Pr(L_n|L_{-n}, \{\lambda_{ijk}\}) \propto \Pr(L_n\cup L_{-n})
\]
We propose to use MCMC to sample from the posterior
distribution of $L_n$.

\textbf{Method 1.} First, sample a starting state $s_0$ from a
Bernoulli distribution with probabilities $(\pi_0, \pi_1)$.  Then,
propose a number $K$ of jumps from a Poisson distribution with rate
parameter $\lambda = \frac{1}{8}\sum\lambda_{ijk}$, which is chosen to
approximate average mutation rate among different contexts. Given $K$,
sample jumping times uniformly on the time interval $(0, t)$,
\textit{i.e.}  from the Dirichlet distribution with concentration
parameters all equal to 1. Therefore, the probability (density) of
proposing a specific path $L' = \{s_0, K, \{t_k\}_{k=1}^K, t\}$ is
\[
q(L') = \pi_{s_0} \frac{\lambda^K \exp(-\lambda)}{K!} \frac{1}{(K-1)!}.
\]
This is an \textit{independence sampler}. Suppose the current guess of
path at position $n$ is $L_n$, we accept the move to the proposed path
$L'$ with probability
\begin{equation}\label{eqn:rejection}
\alpha(L') = \min\{\frac{\pi(L', L_{-n})/q(L')}{\pi(L_n, L_{-n})/q(L_n)}, 1\},
\end{equation}
where $\pi()$ is the complete data likelihood function (\ref{eqn:lik}).

\textbf{Method 2.} The previous method proposes paths from an
approximately uniform distribution. This may be a very inefficient
(high rejection rate) sampling process, although
\citetext{jensen2000probabilistic} used a sampling procedure in the
same spirit. If we use more information from the paths at neighboring
sites, we may be able to improve the efficiency of sampling. The
neighboring paths $L_{n-1}$ and $L_{n+1}$ fragments the evolutionary
time interval $(0,t)$ into smaller segments, within each of which the
states at positions $n-1$ and $n+1$ stay unchanged.

\begin{itemize}
\item Collect the time intervals from site $n-1$ and site $n+1$, so
  that within each of the intervals the states of the neighboring
  sites are unchanged. Let the intervals be represented by a sorted
  array $\{t_0, t_1, \cdots, t_M\}$.
\item For $m = 1, \ldots, M$:
  \begin{itemize}
  \item Let $X$ be a random variable from an exponential
    distribution, representing the jumping time of the middle site
    given that the states of its two neighbors are constant.  For a time
    interval $[t_{m-1}, t_m]$ during which the neighboring sites' states
    are unchanged, suppose $X\sim \text{Exp}(\lambda_{ijk})$, where $i$
    and $k$ are the states of the two neighboring sites in this time
    interval, and $j$ is the starting state of position $n$.
  \item Let $t = t_{m-1}$, and a sample value of $X=x$.
  \item While $t +x < t_m$:
    \begin{enumerate}
    \item[(1)] Add 1 to the number of jumps $k_n \leftarrow k_n +1$.
    \item[(2)] Update $T_n \leftarrow T_n\cup\{t+x\}$.
    \item[(3)] Update $t \leftarrow  t+x$,
    \item[(4)] Sample another value of $X=x$ from $\text{Exp}(\lambda_{ijk})$.
    \end{enumerate}
  \end{itemize}
\end{itemize}

The proposal probability density $q()$ is thus the product of the
appropriate Exponential distribution probability densities for the
holding times at position $n$. The proposal distribution is
independent of any current guess of the path. Therefore, this is also
an \textit{independence sampler}. The rejection rule stays the same,
as in (\ref{eqn:rejection}).


\textbf{Additional note:} A jump at time $\tau$ in the path $L_{n}$
will affect the probability of all jumps that occur after time $\tau$
regardless of their genomic position, because it affects
$\{c_{ijk}(t)\}$ for any $t > \tau$. Therefore, we can't cancel out
factors in the likelihood function. However, we may try to make things
easy by only considering the evolutionary paths at positions $n-2$,
$n-1$, $n+1$, and $n+2$.


\section{Posterior distribution of a path given two neighboring
paths in a tree structure with fixed leaf data}

Suppose we are given a
tree structure representing the evolutionary relationship between
ancestral and extant species. Let $(u,v)$, $(v,a)$ $(v, b)$ be three
branches in the phylogenetic tree.  Suppose we are given the
evolutionary paths on the entire tree for all positions, except for
the paths along these 3 branches at position $n$. In other words, the
states at position $n$ of species $u$, $a$, $b$ are known, but how the
state at position $n$ evolved from $u$ and diverge into states at $a$
and $b$ through their last common ancestor $v$ is unknown. How do we
sample from the posterior distribution of $L_n|v$, which is the path
at position $n$ restricted to branches with one end as node $v$ and
the other end with known states.

\citetext{hobolth2009simulation} reviewed and compared three
approaches to sample paths of discrete-state continuous-time Markov
chain conditional on end-point states, namely the rejection sampling,
direct sampling and uniformization. Method 1 in the previous
section is similar to uniformization, since the neighboring sites may
have state changes, and the rate of jumps depends on the contemporary
states of the neighboring sites. We proposed to use an averaged rate
to for the Poisson distribution, and let the acceptance probability to
do most of the work of correcting the proposal distribution towards the
posterior distribution. Method 2 is forward sampling, which is the
basis of rejection sampling.

\textbf{Rejection sampling} Do forward sampling as described in Method
2 for position $n$ along the branch $(u,v)$ and $(v, a)$. Reject the
path unless the end state agree with the given state at node $a$ at
position $n$.  Then do forward sampling along the branch $(v, b)$,
with the accepted simulated state of node $v$. If the end state does
not agree with the given state at node $b$ at position $n$, start over
the sampling from the beginning, \textit{i.e.} start over from node
$u$. Because we are not taking site $n-2$ and $n+2$ into
consideration, we are not directly sampling from the posterior
distribution of the paths, therefore the rejection sampling procedure
is a way to propose viable paths, then we use the MCMC acceptance rule
(\ref{eqn:rejection}) to reshape the proposal distribution to
the posterior distribution.

A couple of performance statistics should be measured for this
proposed sampling method. These include (1) the success rate for
generating valid paths that satisfy the end conditions; (2) the
acceptance rate of proposed valid paths in MCMC sampling.



\bibliographystyle{namedplus}
\bibliography{biblio}

\end{document}
